zip-0222.rst

```diff
@@ -0,0 +1,322 @@
+  element in both fields MUST have the same ``type``.
+
+- Non-coinbase transactions MUST have at least one of the following:
+  - nonempty transparent inputs
+  - nonempty shielded inputs
+  - nonempty ``tze_inputs``
+
+- Transactions MUST have at least one of the following:
+  - nonempty transparent outputs
+  - nonempty shielded outputs
+  - nonempty ``tze_outputs``
+
+- All ``amount`` field values of ``tze_output`` records MUST be nonnegative and not greater
+  than ``MAX_MONEY``.
+
+- The sum of amounts associated with all transparent ouputs and TZE outputs within 
+  a transaction MUST NOT exceed the sum of the amounts associated with transaction inputs
+  plus the sapling ``valueBalance`` amount.
+
+Changes to signatures over transaction digests
+----------------------------------------------
+
+The newly added parts of the transaction will be included in transaction 
+digests for signatures. See [#zip-0244]_ for the specification of the new
+digests.
+
+Rationale
+=========
+
+Transactions that have both TZE inputs and outputs are required to use a single extension
+type, in order to prevent cross-protocol attacks. The downside is that this prevents all
+TZE-encumbered value from being spent directly into a different TZE type; the value needs
+to go through a regular address in between. This restriction might be relaxed in future
+ZIPs for specific combinations of ``(type, mode)`` pairs that have been analyzed for
+cross-protocol attacks, but we opt here for a fail-safe default behaviour.
+
+Transactions with TZE inputs which do not contain TZE outputs are not subject
+to single-extension or single-mode restrictions; likewise, transactions which
+contain TZE outputs without any TZE inputs may produce TZE outputs for multiple
+extension-type/mode pairs as the potential for cross-protocol attacks in this
+situation is negligible.
+
+An earlier draft version of this ZIP stored the payloads inside transparent inputs and
+outputs. Although this had the advantage of not requiring a transaction format change,
+the consensus rules were significantly more complicated, and the design coupled the
+extension logic too tightly to the transparent address logic. Instead, this ZIP uses
+dedicated transaction fields, and a separate unspent output set.
+
+
+Security and Privacy Considerations
+===================================
+
+This ZIP assumes that the base transaction format is non-malleable. However, the
+``precondition`` and ``witness`` byte sequences are treated here as opaque. It is the
+responsibility of ``tze_verify`` to enforce the following:
+
+- ``precondition`` MUST be non-malleable: any malleation MUST cause ``tze_verify`` to
+  return ``false``.
+- The output of ``tze_verify(mode, precondition, witness, context)`` MUST be deterministic.
+
+ZIPs defining new extension types MUST include a section explaining how any potential
+sources of malleability are handled.
```

Comment:
For ZIPs that dictate how future ZIPs MUST be written on related topics, we should probably have these rules called out in a separate section so they aren't missed.

Reply:
This whole section is essentially that, so I'll indicate it in the title.